File: src/rules.ts
import { existsSync, readFileSync, watch, type FSWatcher } from 'node:fs';
import { join } from 'node:path';
import { parse as parseYaml } from 'yaml';

export interface BlacklistEntry {
  transport: string;
  action: string;
}

export interface RedactionRule {
  field: string;
  /** Regex pattern string. If omitted, the field is dropped entirely. */
  pattern?: string;
}

export interface FloorReservation {
  /** Match condition, e.g. { deadline_within_hours: 72 } */
  match: Record<string, unknown>;
  /** Number of queue slots to reserve for matching candidates. */
  slots: number;
}

export interface QueueConfig {
  target_depth: number;
  low_water_mark: number;
  /** Minimum cluster size to trigger a batched card. Defaults to 3. */
  batch_threshold: number;
  /** Number of queue positions reserved for exploration (high-uncertainty candidates). Defaults to 1. */
  exploration_slots: number;
}

export interface ReversibilityDecl {
  action: string;
  reversible: boolean;
}

export interface VerifierConfig {
  /** How often the post-hoc verifier runs, in minutes. Defaults to 60. */
  interval_minutes: number;
}

export interface PromotionConfig {
  /** Number of approved actions needed before proposing a rule. Defaults to 5. */
  threshold: number;
  /** Minutes to wait after a rejection before re-proposing. Defaults to 1440 (24h). */
  cooldown_minutes: number;
  /** How often the promoter runs, in minutes. Defaults to 120. */
  interval_minutes: number;
}

export interface CredentialScopeDecl {
  action: string;
  refs: string[];
}

export interface Rules {
  blacklist: BlacklistEntry[];
  redaction: RedactionRule[];
  queue: QueueConfig;
  urgent_senders: string[];
  floor: FloorReservation[];
  reversibility: ReversibilityDecl[];
  credential_scopes: CredentialScopeDecl[];
  verifier: VerifierConfig;
  promotion: PromotionConfig;
}

const DEFAULT_QUEUE: QueueConfig = { target_depth: 5, low_water_mark: 2, batch_threshold: 3, exploration_slots: 1 };

const DEFAULT_VERIFIER: VerifierConfig = { interval_minutes: 60 };

const DEFAULT_PROMOTION: PromotionConfig = { threshold: 5, cooldown_minutes: 1440, interval_minutes: 120 };

const EMPTY_RULES: Rules = {
  blacklist: [],
  redaction: [],
  queue: { ...DEFAULT_QUEUE },
  urgent_senders: [],
  floor: [],
  reversibility: [],
  credential_scopes: [],
  verifier: { ...DEFAULT_VERIFIER },
  promotion: { ...DEFAULT_PROMOTION },
};

function loadFile(path: string): Record<string, unknown> | null {
  if (!existsSync(path)) return null;
  const content = readFileSync(path, 'utf8').trim();
  if (!content) return null;
  return parseYaml(content) as Record<string, unknown>;
}

export function loadRules(dir: string): Rules {
  const principles = loadFile(join(dir, 'principles.md'));
  // gmail.md loaded for future use; not parsed into rules yet
  loadFile(join(dir, 'gmail.md'));

  if (!principles) return { ...EMPTY_RULES };

  const blacklist: BlacklistEntry[] = Array.isArray(principles.blacklist)
    ? principles.blacklist.map((e: Record<string, string>) => ({
        transport: e.transport,
        action: e.action,
      }))
    : [];

  const redaction: RedactionRule[] = Array.isArray(principles.redaction)
    ? principles.redaction.map((e: Record<string, string>) => {
        const rule: RedactionRule = { field: e.field };
        if (e.pattern) rule.pattern = e.pattern;
        return rule;
      })
    : [];

  const queueRaw = principles.queue as Record<string, unknown> | undefined;
  const queue: QueueConfig = {
    target_depth: typeof queueRaw?.target_depth === 'number' ? queueRaw.target_depth : DEFAULT_QUEUE.target_depth,
    low_water_mark: typeof queueRaw?.low_water_mark === 'number' ? queueRaw.low_water_mark : DEFAULT_QUEUE.low_water_mark,
    batch_threshold: typeof queueRaw?.batch_threshold === 'number' ? queueRaw.batch_threshold : DEFAULT_QUEUE.batch_threshold,
    exploration_slots: typeof queueRaw?.exploration_slots === 'number' ? queueRaw.exploration_slots : DEFAULT_QUEUE.exploration_slots,
  };

  const urgent_senders: string[] = Array.isArray(principles.urgent_senders)
    ? (principles.urgent_senders as string[]).map((s) => s.toLowerCase())
    : [];

  const floor: FloorReservation[] = Array.isArray(principles.floor)
    ? (principles.floor as Array<Record<string, unknown>>).map((e) => ({
        match: (e.match ?? {}) as Record<string, unknown>,
        slots: typeof e.slots === 'number' ? e.slots : 1,
      }))
    : [];

  const reversibility: ReversibilityDecl[] = Array.isArray(principles.reversibility)
    ? (principles.reversibility as Array<Record<string, unknown>>).map((e) => ({
        action: String(e.action),
        reversible: e.reversible === true,
      }))
    : [];

  const verifierRaw = principles.verifier as Record<string, unknown> | undefined;
  const verifier: VerifierConfig = {
    interval_minutes: typeof verifierRaw?.interval_minutes === 'number' ? verifierRaw.interval_minutes : DEFAULT_VERIFIER.interval_minutes,
  };

  const promotionRaw = principles.promotion as Record<string, unknown> | undefined;
  const promotion: PromotionConfig = {
    threshold: typeof promotionRaw?.threshold === 'number' ? promotionRaw.threshold : DEFAULT_PROMOTION.threshold,
    cooldown_minutes: typeof promotionRaw?.cooldown_minutes === 'number' ? promotionRaw.cooldown_minutes : DEFAULT_PROMOTION.cooldown_minutes,
    interval_minutes: typeof promotionRaw?.interval_minutes === 'number' ? promotionRaw.interval_minutes : DEFAULT_PROMOTION.interval_minutes,
  };

  const credential_scopes: CredentialScopeDecl[] = Array.isArray(principles.credential_scopes)
    ? (principles.credential_scopes as Array<Record<string, unknown>>).map((e) => ({
        action: String(e.action),
        refs: Array.isArray(e.refs) ? (e.refs as string[]).map(String) : [],
      }))
    : [];

  return { blacklist, redaction, queue, urgent_senders, floor, reversibility, credential_scopes, verifier, promotion };
}

export function watchRules(
  dir: string,
  onChange: (rules: Rules) => void,
): { stop: () => void } {
  const watchers: FSWatcher[] = [];

  const reload = () => {
    try {
      onChange(loadRules(dir));
    } catch {
      // Ignore parse errors during mid-write
    }
  };

  const WATCHED = new Set(['principles.md', 'gmail.md']);
  const w = watch(dir, { persistent: false }, (_event, filename) => {
    if (filename && WATCHED.has(filename)) reload();
  });
  watchers.push(w);

  return {
    stop: () => watchers.forEach((w) => w.close()),
  };
}

File: python/steward/rules.py
"""Rules loader. Parses principles.md as YAML. Supports file watching via a poll thread."""
from __future__ import annotations

import threading
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Callable, TypedDict

import yaml


class BlacklistEntry(TypedDict):
    transport: str
    action: str


class RedactionRule(TypedDict, total=False):
    field: str
    pattern: str


@dataclass
class FloorReservation:
    match: dict[str, Any]
    slots: int


@dataclass
class QueueConfig:
    target_depth: int = 5
    low_water_mark: int = 2
    batch_threshold: int = 3
    exploration_slots: int = 1


@dataclass
class ReversibilityDecl:
    action: str
    reversible: bool


@dataclass
class VerifierConfig:
    interval_minutes: int = 60


@dataclass
class PromotionConfig:
    threshold: int = 5
    cooldown_minutes: int = 1440
    interval_minutes: int = 120


@dataclass
class CredentialScopeDecl:
    action: str
    refs: list[str]


@dataclass
class SpendingLimits:
    """Per-payment-class limits enforced by the executor before dispatch.
    Amounts are minor units (pence). None means no limit (issuer-side limit
    is still active as a second defence layer)."""
    max_per_charge_pence: int | None = None
    max_per_day_pence: int | None = None
    max_per_week_pence: int | None = None


@dataclass
class Rules:
    blacklist: list[BlacklistEntry] = field(default_factory=list)
    redaction: list[RedactionRule] = field(default_factory=list)
    queue: QueueConfig = field(default_factory=QueueConfig)
    urgent_senders: list[str] = field(default_factory=list)
    floor: list[FloorReservation] = field(default_factory=list)
    reversibility: list[ReversibilityDecl] = field(default_factory=list)
    credential_scopes: list[CredentialScopeDecl] = field(default_factory=list)
    verifier: VerifierConfig = field(default_factory=VerifierConfig)
    promotion: PromotionConfig = field(default_factory=PromotionConfig)
    spending_limits: SpendingLimits = field(default_factory=SpendingLimits)


def _load_file(path: Path) -> dict[str, Any] | None:
    if not path.exists():
        return None
    content = path.read_text(encoding="utf-8").strip()
    if not content:
        return None
    parsed = yaml.safe_load(content)
    return parsed if isinstance(parsed, dict) else None


def load_rules(directory: str | Path) -> Rules:
    d = Path(directory)
    principles = _load_file(d / "principles.md")
    _load_file(d / "gmail.md")  # reserved for future per-surface rules

    if not principles:
        return Rules()

    blacklist: list[BlacklistEntry] = []
    for entry in principles.get("blacklist") or []:
        blacklist.append({"transport": entry["transport"], "action": entry["action"]})

    redaction: list[RedactionRule] = []
    for entry in principles.get("redaction") or []:
        rule: RedactionRule = {"field": entry["field"]}
        if entry.get("pattern"):
            rule["pattern"] = entry["pattern"]
        redaction.append(rule)

    queue_raw = principles.get("queue") or {}
    queue = QueueConfig(
        target_depth=queue_raw.get("target_depth", 5),
        low_water_mark=queue_raw.get("low_water_mark", 2),
        batch_threshold=queue_raw.get("batch_threshold", 3),
        exploration_slots=queue_raw.get("exploration_slots", 1),
    )

    urgent_senders = [str(s).lower() for s in (principles.get("urgent_senders") or [])]

    floor: list[FloorReservation] = []
    for entry in principles.get("floor") or []:
        floor.append(
            FloorReservation(
                match=entry.get("match") or {},
                slots=int(entry.get("slots", 1)),
            )
        )

    reversibility: list[ReversibilityDecl] = []
    for entry in principles.get("reversibility") or []:
        reversibility.append(
            ReversibilityDecl(
                action=str(entry.get("action")),
                reversible=entry.get("reversible") is True,
            )
        )

    verifier_raw = principles.get("verifier") or {}
    verifier = VerifierConfig(interval_minutes=verifier_raw.get("interval_minutes", 60))

    promotion_raw = principles.get("promotion") or {}
    promotion = PromotionConfig(
        threshold=promotion_raw.get("threshold", 5),
        cooldown_minutes=promotion_raw.get("cooldown_minutes", 1440),
        interval_minutes=promotion_raw.get("interval_minutes", 120),
    )

    credential_scopes: list[CredentialScopeDecl] = []
    for entry in principles.get("credential_scopes") or []:
        credential_scopes.append(
            CredentialScopeDecl(
                action=str(entry.get("action")),
                refs=[str(r) for r in (entry.get("refs") or [])],
            )
        )

    spending_raw = principles.get("spending_limits") or {}
    spending_limits = SpendingLimits(
        max_per_charge_pence=spending_raw.get("max_per_charge_pence"),
        max_per_day_pence=spending_raw.get("max_per_day_pence"),
        max_per_week_pence=spending_raw.get("max_per_week_pence"),
    )

    return Rules(
        blacklist=blacklist,
        redaction=redaction,
        queue=queue,
        urgent_senders=urgent_senders,
        floor=floor,
        reversibility=reversibility,
        credential_scopes=credential_scopes,
        verifier=verifier,
        promotion=promotion,
        spending_limits=spending_limits,
    )


@dataclass
class RulesWatcher:
    stop: Callable[[], None]


def watch_rules(
    directory: str | Path,
    on_change: Callable[[Rules], None],
    poll_interval: float = 1.0,
) -> RulesWatcher:
    """Simple polling watcher — reloads when principles.md or gmail.md mtime changes."""
    d = Path(directory)
    watched = [d / "principles.md", d / "gmail.md"]
    stop_event = threading.Event()

    def snapshot() -> tuple[float, ...]:
        return tuple(p.stat().st_mtime if p.exists() else 0.0 for p in watched)

    last = snapshot()

    def loop() -> None:
        nonlocal last
        while not stop_event.wait(poll_interval):
            current = snapshot()
            if current != last:
                last = current
                try:
                    on_change(load_rules(d))
                except Exception:
                    pass

    thread = threading.Thread(target=loop, daemon=True)
    thread.start()
    return RulesWatcher(stop=stop_event.set)
